### ISSUE TITLE
How does OctConv really implement strided convolutions

### ISSUE THREAD

**Issue by andravin** [terrychenism/OctaveConv#4]
https://github.com/terrychenism/OctaveConv/blob/5fa78aeb57ddd595057c56b8650a514019b63da9/symbol_octConv.py#L46

**It is not clear to me from reading the paper how OctConv implements strided convolutions. Because the paper is not explicit about this, it would be great if we could get clarification from the authors.**

This also leads to a deeper conversation about how OctConv is handling downsampling / upsampling, which maybe should be had here, because this github repo is the most popular implementation.

Seems most readers assume that OctConv implements strided convolution the same way that the downsampling operation is implemented, using average pooling followed by regular convolution. This might be true, but I cannot see where the paper actually says this.

Adding to the confusion, I think this premise from the paper is wrong:

> shifting the location by half step is to ensure the down-sampled maps well aligned with the input

One reason the downsampled maps would be shifted by a half step relative to the full resolution input would be if they were produced by convolution with an even-sized filter, such as 2x2 average pooling with stride 2x2. If instead the downsampled maps were produced by convolution with an odd sized filter, such as a 3x3 convolution with padding = 1, then the output would be perfectly aligned with the input.

Another cause of feature misalignment is "valid" padding, in other words, when padding is zero, which shifts the alignment of the outputs relative to the inputs. This is actually the cause of misalignment corrected by the "Deformable Convolutional Neural Networks" paper: [1]

> In the original Inception-ResNet [51] architecture, multiple layers of valid convolution/pooling are utilized, which brings feature alignment issues for dense prediction tasks. 

[...]

> Aligned- Inception-ResNet does not have the feature alignment problem, by proper padding in convolutional and pooling layers

That is actually the citation used by the OctConv paper to make this claim:

> we adopt average pooling to get more accurate approximation. This helps alleviate misalignments that appear when aggregating information from different scales

So that appears to be a misattribution, because the "Deformable Convnets" paper was actually addressing an alignment problem caused by valid convolutions used by Inception-Resnet.

Figure 3 from the paper is misleading. The strided convolution with 3x3 filter does not cause misalignment. It is the subsequent up-sampling using nearest neighbor interpolation that causes it. Nearest neighbor is equivalent to upsampling with the filter [1 1] which has an even number of taps. If the upsampling instead used a 3-tap or 5-tap interpolation filter, the upsampled feature map would remain aligned with the original.

This is the reason that Laplacian Pyramids usually use odd-sized filters, to keep the multiple layers aligned.

There are good engineering reason for using the simple 2-tap filters (average pooling and nearest neighbor interpolation). They are fast, they are already implemented by major frameworks, they do not require any padding. So I am not questioning these design choices, just the justification for them.

References:
[1] Dai, Jifeng, et al. "Deformable Convolutional Networks." arXiv preprint arXiv:1703.06211 (2017).
https://arxiv.org/abs/1703.06211

**Comment by deyituo** [comment]
I searched oct conv in github and found that the implementations are the same as this one. It seems like that the strided convolutional and upsample/downsample should be considered carefullly when the size is odd. @andravin

**Comment by andravin** [comment]
I agree that we have to be careful when the strided convolution (or max-pool) filter size is odd, because the OctConv downsampling/upsampling filter size is even.

Just passing the stride to the HH, LL, HL, and LL convolutions inside OctConv would be the wrong thing to do (and at least 1 github implementation does this), because it does not correct the half-pixel shift misalignment that is caused by downsampling with an even sized filter. 

Average pooling is the 2-tap filter `[1 1]/2`, and nearest neighbor upsampling is the filter `[1 1]`. Average pooling shifts the feature map by 1/2 pixel, and, with proper padding, nearest neighbor upsampling unshifts the feature map by the same amount.

Of course, the authors were not required to choose even-size downsampling/upsampling filters. **Laplacian pyramids almost always use odd-size filters so that there are no half-pixel phase shifts between layers.**

There are still different choices you could make when implementing strided OctConv, and we could discuss each of them. But we still would not know exactly what the authors did and would not be confident that we were reproducing the paper accurately.

**Comment by savourylie** [comment]
I believe it's stated in this part of the paper,

> However, since the index of X_H can only be an integer, we could either round the index to (2∗p+i, 2∗q+j) or approximate the value at (2∗p+0.5+i, 2∗q+0.5+j) by averaging all 4 adjacent locations. The first one is also known as strided convolution and the second one as average pooling. As we discuss in Section 3.3 and Fig. 3, strided convolution leads to misalignment; we therefore use average pooling to approximate this value for the rest of the paper.

that average pooling is chosen over strided convolution **for the rest of the paper**. Please also see https://github.com/ThoroughImages/OctConv/issues/1#issuecomment-486040529

**Comment by andravin** [comment]
That does not actually say how they port `Conv2d(stride=2)` to `OctConv(stride=2)`, which is the subject of this issue.

It is true however that mixing `Conv2d(kernel_size=3, stride=2)` with `AvgPool(kernel_size=2, stride=2)` is not good, because this creates 2 different half-scale grids that are shifted relative to each other.

So the authors probably did replace Conv2d(stride=2) with something like `AvgPool(stride=2)->Conv2d(stride=1`), as this repo does, but the exact formulation is not documented in the paper.

**Comment by DevinXYang** [comment]
Hello, How do you handle with feature map size is 7x7?
When feature map size is 7x7, using `Pooling(data, (2, 2), 'avg', stride=(2, 2))` would make it to 3x3, and using `UpSampling(avg_pool, scale=2, sample_type='nearest', num_args=1)` would make it to 6x6 but not 7x7?How do you handle this?
I read your code. it seem that you don't use OctConv in last residual Block?

**Comment by andravin** [comment]
The authors released their reference implementation: https://github.com/facebookresearch/OctConv

**Comment by andravin** [comment]
@PistonY Looks like the authors just set `alpha=0` in the last ResNet stages, so that the 7x7 feature map is not downsampled to a lower resolution:

```
        # ratio is forced to be 0. for the last stage
        # (because do 3x3 conv on 3.5x3.5 resolution map does not make sense)
```

https://github.com/facebookresearch/OctConv/blob/253139b1dc842f04030077d983b2cabb5a754b3a/utils/symbol/symbol_resnetv2.py#L82-L83

**Comment by DevinXYang** [comment]
@andravin Yeah,thanks.So it's not a totally plug-and-play "tools" and not suitable for singular side length featuer maps.

**Comment by andravin** [comment]
@PistonY I guess .. in my experiments, I avoided this issue by training with 256x256 input image resolution, so all layer sizes are a power of 2. If the filter was adapted to take the image boundary into account, one could downsample from 7x7 to 4x4. Or maybe it would make more sense to downsample the last 14x14 layer to 8x8.

**Comment by andravin** [comment]
Closing this issue, because we can now look at the reference implementation to see how OctConv implements strided convolutions: https://github.com/facebookresearch/OctConv

**Comment by jackchinor** [comment]
> Hello, How do you handle with feature map size is 7x7?
> When feature map size is 7x7, using `Pooling(data, (2, 2), 'avg', stride=(2, 2))` would make it to 3x3, and using `UpSampling(avg_pool, scale=2, sample_type='nearest', num_args=1)` would make it to 6x6 but not 7x7?How do you handle this?
> I read your code. it seem that you don't use OctConv in last residual Block?

have you solved this problem? I met the same problem with you when deal with last octave conv layer. I don't know how to handle it.

**Comment by jackchinor** [comment]
> Hello, How do you handle with feature map size is 7x7?
> When feature map size is 7x7, using `Pooling(data, (2, 2), 'avg', stride=(2, 2))` would make it to 3x3, and using `UpSampling(avg_pool, scale=2, sample_type='nearest', num_args=1)` would make it to 6x6 but not 7x7?How do you handle this?
> I read your code. it seem that you don't use OctConv in last residual Block?

@PistonY have you solved this problem? I met the same problem with you when deal with last octave conv layer. I don't know how to handle it.
